Proofread, reorganized, and added to all the "Python on Windows" guides (#794)
discord-python/site: pydis_site/apps/content/resources/guides/python-guides/windows/common-issues.md

and click on "Manage app execution aliases". In the list you should see two options with title "App Installer" and
descriptions `python.exe` and `python3.exe`. Disable both of them.

Note that if you are following instructions telling you to run a command starting with `python3`, those instructions are
intended for Unix/macOS systems. Try just using `python` instead.

If after doing this you have an issue with the Python command not being detected, see the question below.

### When I try and run my code with `python` in the terminal I get an error saying the command was not recognized

If you have not installed Python, you will need to do that. Follow our guide [here](../installing-and-using-python).

If you have installed python and are still having the issue, it is likely that you didn't check the "Add python.exe to
PATH" checkbox on the first page of the installer. See [our guide on adding Python to PATH](../putting-python-on-path)
for how to fix this. You can also use the `py` launcher instead of `python` by just replacing `python` in your command
with `py`. See [this guide](../installing-and-using-python/#the-py-launcher) for more information on that.

> ### I `pip` installed a package but when running my code get a `ModuleNotFoundError`

Reviewer: ```suggestion
### I `pip` installed a package but when running my code I get a `ModuleNotFoundError`
```

-   #### Are you actually getting a `ModuleNotFoundError`?

    If you are using a code editor such as VSCode or PyCharm you may get a squiggly line under your import saying the
    module couldn't be found, it is possible that this is just an mistake by the editor, so try actually running your

>     code with to ensure if it actually errors.

Reviewer: ```suggestion
so try to actually run your code the way it is to ensure that it actually errors.
```

    If your code runs fine, you could try restarting your editor. If you have newly installed a module it may just not

>     have detected it yet. If that doesn't help, you may need to configure your editor to ensure it is looking for the

Reviewer: Actually, some editors can still show that error even if you're showing the correct virtual environment with the installed package.
I think it was opencv IIRC.

Author: Is that this https://stackoverflow.com/questions/34365044/pycharm-does-not-recognize-cv2-as-a-module ?
Where the fix is something like `from cv2 import cv2`. Could add a note to that SO link.

        import has to be `import dotenv`
    -   [**Pillow**](https://pypi.org/project/Pillow/): You need to `pip install Pillow`, but the import is
        usually `from PIL import Image`.

    Normal imports in Python can't have dashes (`-`) in them as it would be interpreted as subtraction. Though [there
    are workarounds](https://stackoverflow.com/questions/7583652/python-module-with-a-dash-or-hyphen-in-its-name).

-   #### Are you installing to the same environment you're running your code from?

    This is a very common issue. When you install a module from PyPI, you will install it into a single Python
    environment, the one that the `pip` you invoked is part of. This could be different to the one you are running the
    script from if it's using a different Python installation, or a virtual environment.

    ##### I'm using PyCharm

    > PyCharm creates a virtual environment for each project by default, so if you're installing a module using Pycharm
    > you need to make sure you also run your code through PyCharm. This also applies the other way round, if you want
    > to use a module in PyCharm you need to install it through PyCharm.
    >

>     > It's also possible to change PyCharm to use your system environment, see [their guide on configuring your

Reviewer: We can add a section here that helps people identify the currently active virtual environment in PyCharm.
If you look at the bottom right, you will see your project root's name between parenthesis along with the python version like this.
![image](https://user-images.githubusercontent.com/48383734/200617946-ad54c699-b29c-4582-a404-5d9efb7d8848.png)

If you click on that, you can see the full path to your virtual env here
![image](https://user-images.githubusercontent.com/48383734/200618479-1306991f-ecdb-4a01-8a57-e82b54719838.png)

So if you want to install a package into that one from the terminal in Pycharm, you might want to make sure that you have the correct one active by checking the venv name between parenthesis here
![image](https://user-images.githubusercontent.com/48383734/200618764-48e1a10f-d313-4fcd-96fd-96b70e62f828.png)

I remember making this mistake when I was first starting

Author: Sure. I added some images and info mentioning these.

    > interpreter](https://www.jetbrains.com/help/pycharm/configuring-python-interpreter.html) for more information.

    ##### I installed the module using pip from the command line

>     > Run `pip -V` and look at the path returned it should be in the form `<PATH_TO_PYTHON>\Lib\site-packages\pip`.

Reviewer: ```suggestion
    > Run `pip -V` and look at the path it returned, which should be in the form `<PATH_TO_PYTHON>\Lib\site-packages\pip`.
```

    PATH](../putting-python-on-path/#advanced-method-manually-edit-the-path) for how to verify and fix this.

Projects that add a command line program usually work by adding an executable to the `\Scripts` folder of your Python
install (the full path of which is `C:\Users\<user>\AppData\Local\Programs\Python\Python311\Scripts` in a default 3.11
install). If you checked "Add python.exe to PATH" when installing Python, this folder will have been added to PATH, so
anything there should be runnable directly from the command line. Virtual environments also have their own `\Scripts`
directories that things can be installed to and are effectively on the PATH when the venv is properly activated.

### I get a `SyntaxError` when trying to run `pip`,`python`, or another command.

When you type `python` in the terminal you enter into the python REPL ([read-evaluate-print
loop](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop)), also known as the Python console. This lets
you run `Python` code line by line without having to create a file. You also may have opened a Python REPL by clicking
on the Python App in your Start Menu.

> General commands like `pip` and `python` should be run in the normal terminal, not in the Python REPL. It usually has

Reviewer: I think this could cause confusion for beginners, because I might ask a question like "normal" means here.
So maybe it'll be better if we be more specific by saying it's the windows terminal, e.g. command prompt

Author: Good point. I'll clarify that "normal" probably means Command Prompt or PowerShell.

---
title: Common Issues Using Python on Windows
description: A list of common issues with Python on Windows and how to fix them
icon: fab fa-windows
toc: 3
---

> ### When I run `python` in the terminal I get no result, or the Windows Store opens!

Reviewer 2: ```suggestion
### When I run `python` in the terminal I get no result, or the Microsoft Store opens!
```
probably my mistake, but since I see it here. I write the wrong thing all the time :P

By default Windows has an alias for `python` in the terminal to guide you to install it from the [Microsoft
Store](https://apps.microsoft.com/store/search/python). We don't recommend installing Python from the Microsoft
Store, see why [here](../microsoft-store). To disable this alias, search "App execution aliases" in Windows search
and click on "Manage app execution aliases". In the list you should see two options with title "App Installer" and

> descriptions `python.exe` and `python3.exe`. Disable both of them.

Reviewer 2: I wrote this, but I sort of wonder whether it would be best not to suggest this. Disabling the alias doesn't really help with anything, as having python on path will overwrite it which is the intended behaviour. 

All this means is that if the user *does* try and use the microsoft store version it wont work.

Author: I think it's important to mention the "Manage app execution alias" setting *somewhere* - I didn't even know about it until writing the guides. (Though it is mentioned in  `/microsoft-store/#confusing-app-execution-alias-behaviour` come to think of it.)

Whether to *suggest* it or not I have no opinion on. Though I believe it does work more or less like a normal path entry as explained here: https://github.com/python-discord/site/pull/794#discussion_r1020826357

Reviewer 2: I see, I didn't realise it was a normal path entry.

I also just realised that if you install python from the ms store it adds the other entries (pip, etc) as app execution aliases (there are loads...)
![image](https://user-images.githubusercontent.com/22353562/201529405-96756802-e0f9-4c87-8414-fa7741fbfb26.png)
![image](https://user-images.githubusercontent.com/22353562/201529431-5c23f048-b01e-4ff1-9ff8-50324919d75a.png)
![image](https://user-images.githubusercontent.com/22353562/201529454-350a3c8e-181f-456c-b434-edd0da356cd0.png)

Are the app installer ones the two that are there by default?

Author: Yeah (at least on Win 10) when you have the App Installer ones on but no MS store Python installed a python.exe and python3.exe still show up in `C:\Users\user\AppData\Local\Microsoft\WindowsApps`.

-   #### Did you use the correct module name?

    Double check that you haven't made a typo in the name you are importing, or in what you installed from
    [PyPI](https://pypi.org/), you need to make sure you type the name exactly as it should be.

    Also, **the name you should `import` may not be the same as the name you `pip install`**. Check the module's
    documentation or PyPI page if you are unsure. Examples of this are:

    -   [**opencv-python**](https://pypi.org/project/opencv-python/): You need to `pip install opencv-python`, but the
        import has to be `import cv2`
    -   [**discord.py**](https://pypi.org/project/discord.py/): You need to `pip install discord.py`, but the import has
        to be `import discord`
    -   [**python-dotenv**](https://pypi.org/project/python-dotenv/): You need to `pip install python-dotenv`, but the
        import has to be `import dotenv`
    -   [**Pillow**](https://pypi.org/project/Pillow/): You need to `pip install Pillow`, but the import is
        usually `from PIL import Image`.

    Normal imports in Python can't have dashes (`-`) in them as it would be interpreted as subtraction. Though [there

>     are workarounds](https://stackoverflow.com/questions/7583652/python-module-with-a-dash-or-hyphen-in-its-name).

Reviewer 2: Is this something you've seen people have an issue with? Is it with their own modules or installed ones?

Author: I think I've seen the issue a few times in a help channel, though can't recall exactly what was imported.

I do feel it's useful to explain why dashes (such a common character to name files/folder/packages with) don't work the same in Python syntax - even in imports.

Reviewer 2: I don't think we should link to the workaround, as I don't think there's a case where that should be used and it will likely cause other problems.

We could mention that if a package from pypi has a hyphen in the name the actual import is probably different, and if they're trying to import a local file they should rename it.

Author: Alright, I can remove the workaround. I agree it's bad to send people towards unrecommended problem fixes.

    >
    > Alternatively, you can pip install to the currently activated environment. If you have configured VSCode to use a
    > virtual environment, it should automatically activate it when you open a new terminal, so running the
    > `pip install` command in the VSCode terminal should work. If you are not using a virtual environment, you can use
    > the `py` launcher to specify the installation you want to install to.

### I `pip` installed a command line program but it isn't recognized

You pip installed a command line program like [`pyinstaller`](https://pypi.org/project/pyinstaller/) or
[`black`](https://pypi.org/project/black/) that is intended to be run from the command line, but it does not work. Two
common causes for this are:

-   You are using Python from the Microsoft Store ([which we don't recommend](../miscrosoft-store)).

    To check if this is the case, type `pip -V`. If the path output includes something like
    `PythonSoftwareFoundation.Python.3.10_3.10.2288.0_x64__qbz5n2kfra8p0`, you are using Python from the Microsoft
    Store. To fix this you will either have to use the full path to the scripts you want to use, or [uninstall the

>     Microsoft Store Python](/static/images/content/python-on-windows/ms_store_uninstall.png) and [install Python

Reviewer 2: would it be better to link to the uninstall section on the page instead here?

Author: Good ideal. I'll make `microsoft-store/#uninstalling` a section and change this link.
